_source/_posts/2022-07-20-express-prisma-tailwind.md

```diff
@@ -0,0 +1,665 @@
+});
+
+passport.deserializeUser((obj, next) => {
+  next(null, obj);
+});
+
+app.use(logger('dev'));
+app.use(express.json());
+app.use(express.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, 'public')));
+
+app.use('/login', passport.authenticate('oidc'));
+
+app.use('/authorization-code/callback',
+  passport.authenticate('oidc', { failureRedirect: '/error' }),
+  (req, res, next) => {
+    res.redirect('/dashboard');
+  }
+);
+
+app.post('/logout', (req, res) => {
+  req.logout();
+  req.session.destroy();
+  res.redirect('/');
+});
+
+app.use('/', indexRouter);
+app.use('/dashboard', dashboardRouter);
+
+// catch 404 and forward to error handler
+app.use(function(req, res, next) {
+  next(createError(404));
+});
+
+// error handler
+app.use(function(err, req, res, next) {
+  // set locals, only providing error in development
+  res.locals.message = err.message;
+  res.locals.error = req.app.get('env') === 'development' ? err : {};
+
+  // render the error page
+  res.status(err.status || 500);
+  res.render('error');
+});
+
+module.exports = app;
+```
+
+You updated your routes to properly use the `dashboard` routes as well as any login or logout routes that Okta will use.  You also set up passport to use Okta using the values the Okta CLI produced in the `.okta.env` file.  
+
+### Add your views to bring your project to life
+
+By default, `express-generator` will add a few views for you.  You will need to edit them and add a new one of your own.  Start by opening the `layout.pug` file in the `views` directory and replacing the code there with the following.
+
+```pug
+doctype html
+html
+  head
+    title= title
+    link(rel='stylesheet', href='/stylesheets/output.css')
+  body
+    nav.flex.items-center.justify-between.flex-wrap.bg-teal-500.p-6.flex.items-center.flex-shrink-0.text-white.mr-6
+      span.font-semibold.text-xl.tracking-tight.pr-2 The Workout Tracker
+      .w-full.block.flex-grow(class='lg:flex lg:items-center lg:w-auto')
+        .text-sm(class='lg:flex-grow')
+        div
+          if(isAuthenticated)
+            form(action="logout" method="POST") 
+              button(type="submit").inline-block.text-sm.px-4.py-2.leading-none.border.rounded.text-white.border-white.mt-4 Logout
+          else 
+            a.inline-block.text-sm.px-4.py-2.leading-none.border.rounded.text-white.border-white.mt-4(href='Login' class='hover:border-transparent hover:text-teal-500 hover:bg-white lg:mt-0') Login
+    
+    block content
+```
+
+You'll notice a couple of things about this file.  First, the classes are very specific and plentiful.  This is typical of Tailwind CSS.  Rather than defining how a specific element looks, Tailwind CSS strives to define how a particular style looks, then you can add and remove styles from elements to fit the desired look.  
+
+The other thing you'll see is the Login/Logout button that will change based on the `isAuthenticated` property.  
```

Comment:
```suggestion
The other thing you'll see is the **Login/Logout** button that will change based on the `isAuthenticated` property.  
```
bolding the text since it refers to a UI element